core/domain/auth_jobs_one_off.py

```diff
@@ -39,6 +39,258 @@
+                # the same email can not be claimed while the deletion request
+                # is pending.
+                disabled=marked_as_deleted)
+        elif marked_as_deleted != assoc_info[cls.FIREBASE_ACCOUNT_DISABLED_KEY]:
+            firebase_admin.auth.update_user(auth_id, disabled=marked_as_deleted)
+
+    @classmethod
+    def report_assocs_missing(cls, assoc_info_pairs):
+        """Yields debug information for each of the given associations.
+
+        NOTE: Since assoc_by_auth_id and Firebase accounts are keyed by auth_id,
+        the only kinds of associations that could ever hit this function are
+        UserAuthDetailsModel.
+
+        Args:
+            assoc_info_pairs: list(tuple(str, (str, bool))). The list of
+                associations that do not correspond to any auth_id.
+
+        Yields:
+            str. A debug string for each association.
+        """
+        for _, (user_id, _) in assoc_info_pairs:
+            yield (
+                'UserAuthDetailsModel(id="%s") does not correspond to a unique '
+                'firebase_auth_id' % user_id)
+
+    @classmethod
+    def report_assoc_collisions(cls, auth_id, assoc_info_pairs):
+        """Yields debug information for associations mapped to the same auth_id.
+
+        NOTE: Since assoc_by_auth_id and Firebase accounts are keyed by auth_id,
+        and are therefore guaranteed to have no collisions, the associations
+        reported will always come from UserAuthDetailsModel.
+
+        Args:
+            auth_id: str. The auth_id to check.
+            assoc_info_pairs: list(tuple(str, (str, bool))). The list of
+                associations that do not correspond to any auth_id.
+
+        Yields:
+            str. A debug string for the associations with collisions.
+        """
+        user_id_collisions = sorted(
+            '"%s"' % user_id for assoc_key, (user_id, _) in assoc_info_pairs
+            if assoc_key == cls.ASSOC_BY_USER_ID_KEY)
+
+        if len(user_id_collisions) > 1:
+            yield '%d UserAuthDetailsModels have auth_id="%s": %s' % (
+                len(user_id_collisions), auth_id, ', '.join(user_id_collisions))
+
+    @classmethod
+    def report_assoc_inconsistencies(
+            cls, auth_id, assoc_by_auth_id=None, assoc_by_user_id=None,
+            firebase_account_disabled=None):
+        """Reports inconsistencies between the given values.
+
+        IMPORTANT: The names of the keyword arguments MUST match the values of
+        their corresponding class keys: ASSOC_BY_AUTH_ID_KEY,
+        ASSOC_BY_USER_ID_KEY, FIREBASE_ACCOUNT_DISABLED_KEY.
+
+        Args:
+            auth_id: str. The auth_id of the user.
+            assoc_by_auth_id: tuple|None. The (user_id, deleted) properties from
+                the UserIdByFirebaseAuthIdModel corresponding to auth_id. NOTE:
+                UserIdByFirebaseAuthIdModel is keyed by auth_id,
+                so this argument will never use the default value of None.
```

Comment:
So why is there a default value specified in the docstring? Suggest dropping it if it can't be None.

Reply:
Because the argument itself is None by-default, so leaving it out would be even more confusing IMO.

Reply:
No, I don't think so -- your docstring actually explains the discrepancy well, and you want the "thing you need to do although you don't want to do it" to be in as few places as possible. If people are confused about it, the first place they'll look at is the docstring.

Reply:
Done